File: mobile/lib/features/auth/domain/validation.dart
// Mirror of web/backend/src/services/validation.ts — keep these in sync.
// Rules: 8+ chars, ≥1 uppercase Latin, ≥1 lowercase Latin, ≥1 digit, only
// Latin letters and common punctuation (no Cyrillic).
final _passwordRe = RegExp(
  r'^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)'
  r"""[A-Za-z0-9!@#\$%^&*()_+\-=\[\]{};':",.<>?/`~|\\]{8,}$""",
);

bool isValidPassword(String value) => _passwordRe.hasMatch(value);

final _emailRe = RegExp(r'^[^\s@]+@[^\s@]+\.[^\s@]+$');

bool isValidEmail(String value) => _emailRe.hasMatch(value);

final _managerCodeRe = RegExp(r'^\d{6}$');

bool isValidManagerCode(String value) => _managerCodeRe.hasMatch(value);

final _otpRe = RegExp(r'^\d{6}$');

bool isValidOtp(String value) => _otpRe.hasMatch(value);

// E.164 international format — mirror of backend `isValidPhone`. The mobile
// `phone_form_field` outputs phones in this canonical form.
final _phoneRe = RegExp(r'^\+\d{10,15}$');

bool isValidPhone(String value) => _phoneRe.hasMatch(value);

/// Trimmed name must be 1..50 chars. Used for first/last name on profile edit.
bool isValidName(String value) {
  final trimmed = value.trim();
  return trimmed.isNotEmpty && trimmed.length <= 50;
}

File: web/frontend/src/lib/validation.ts
// Single source of truth for input validation regexes on the web side.
// MUST stay in sync with web/backend/src/services/validation.ts and
// mobile/lib/features/auth/domain/validation.dart — if you change one,
// change all three.

export const EMAIL_RE = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;

// E.164: leading + and 10–15 digits.
export const PHONE_RE = /^\+\d{10,15}$/;

export const MANAGER_CODE_RE = /^\d{6}$/;

// 8+ chars, ≥1 uppercase Latin, ≥1 lowercase Latin, ≥1 digit, only Latin
// letters and common punctuation.
export const PASSWORD_RE =
  /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)[A-Za-z0-9!@#$%^&*()_+\-=\[\]{};':",.<>?/`~|\\]{8,}$/;

export const isValidEmail = (v: string) => EMAIL_RE.test(v);
export const isValidPhone = (v: string) => PHONE_RE.test(v);
export const isValidManagerCode = (v: string) => MANAGER_CODE_RE.test(v);
export const isValidPassword = (v: string) => PASSWORD_RE.test(v);
